From 7e3f9cd96b6baa8ba3c266941bc697689ce64a73 Mon Sep 17 00:00:00 2001
From: Tamal <tasaha@habana.ai>
Date: Fri, 02 Aug 2024 15:59:48 +0300
Subject: [PATCH] [SW-161727] g3: init bist_skip_enable bit

This patch resets bist_skip_enable bit during
fw_all reset in ppboot stage.

Change-Id: I90a8b825a1adb98b151e876c8205b0602095ceeb
Signed-off-by: Tamal <tasaha@habana.ai>
---

diff --git a/arch/gaudi3/arc/asic.c b/arch/gaudi3/arc/asic.c
index e03f213..e5a9c62 100644
--- a/arch/gaudi3/arc/asic.c
+++ b/arch/gaudi3/arc/asic.c
@@ -218,13 +218,35 @@
 	u64 sram_loadaddr;
 	bool ret;
 
+	cold_rst_data.data = regmap_readl(mmCOLD_RST_DATA);
+
 	/* If FW_ALL reset then skip bist */
 	if (hl_rst_src_get() == HL_FW_ALL_RST) {
 		preboot_puts("bist skipped for 96 MB SRAMs during FW all reset\n");
+
+		/* a) when bist is performed, this bist_skip_enable
+		 * is set to indicate hard reset to be triggred.
+		 * b) hard reset is triggered when bist_skip_enable
+		 * is set after bist. After hard reset, this
+		 * bist_skip_enable is reset again.
+
+		 * Post bist hard reset is disabled as of now, (b)
+		 * is never executed, bist_skip_enable is never reset.
+		 * This leads to problem, when upgrading fw.
+
+		 * To fix the above problem, if FW_ALL reset,
+		 * initialize bist_skip_enable bit.
+		 */
+
+		/* TODO: This fix can be reverted after fw release
+		 * to maintain same code for gaudi2 and gaudi3.
+		 */
+
+		cold_rst_data.bist_skip_enable = 0;
+		regmap_writel(cold_rst_data.data, mmCOLD_RST_DATA);
 		return;
 	}
 
-	cold_rst_data.data = regmap_readl(mmCOLD_RST_DATA);
 	if (cold_rst_data.bist_skip_enable) {
 		preboot_puts("bist skipped for 96 MB SRAM\n");
 		return;
